File: python/zoned_time_lite/working_hours.py
"""Working hours and business day utilities."""

from __future__ import annotations

from datetime import datetime

from . import DEFAULT_TIMEZONE, DEFAULT_WORKING_HOURS, DEFAULT_WORKING_DAYS
from .types import WorkingDays
from .timezone_registry import validate_platform_timezone
from .validator import validate_date, validate_timezone, validate_working_hours, validate_working_days
from .time_converter import to_timezone_parts


def in_working_hours(
    date: datetime,
    tz: str = DEFAULT_TIMEZONE,
    start: str = DEFAULT_WORKING_HOURS["start"],
    end: str = DEFAULT_WORKING_HOURS["end"],
) -> bool:
    """
    Check if a timestamp falls within working hours for a given timezone.

    Converts the input timestamp to the specified timezone's local time and checks
    if it falls within the defined working hours range. Handles timezone-aware 
    calculations correctly, including DST transitions.

    Args:
        date: The date to check (must be a valid datetime object)
        tz: IANA timezone identifier (defaults to 'Europe/London')
        start: Start time in HH:MM format (defaults to '09:00')
        end: End time in HH:MM format (defaults to '17:30')

    Returns:
        True if the timestamp is within working hours in the specified timezone

    Raises:
        ValueError: If date is invalid (NaT) or outside supported range (1970-2100)
        ValueError: If timezone is not supported or unavailable on platform
        ValueError: If time format is invalid (must be HH:MM format, e.g., '09:00')

    Example:
        >>> from datetime import datetime, timezone
        >>> utc_time = datetime(2024, 7, 15, 13, 0, 0, tzinfo=timezone.utc)  # 1 PM UTC
        >>> 
        >>> # Default working hours (09:00-17:30)
        >>> in_working_hours(utc_time, 'Europe/London')     # True (14:00 BST)
        >>> in_working_hours(utc_time, 'America/New_York')  # False (09:00 EDT, at start)
        >>> in_working_hours(utc_time, 'Asia/Tokyo')        # False (22:00 JST, after hours)
        >>> 
        >>> # Custom working hours
        >>> in_working_hours(utc_time, 'Europe/London', '08:00', '16:00')  # True
        >>> 
        >>> # Midnight-spanning hours (22:00-06:00)
        >>> night_time = datetime(2024, 7, 15, 23, 0, 0, tzinfo=timezone.utc)  # 11 PM UTC
        >>> in_working_hours(night_time, 'Europe/London', '22:00', '06:00')  # True (00:00 BST)
    """
    validate_date(date)
    validate_timezone(tz)
    validate_working_hours(start, end)
    validate_platform_timezone(tz)

    # Get timezone-local time components
    local_parts = to_timezone_parts(date, tz)

    # Parse start and end times
    start_hour, start_minute = map(int, start.split(":"))
    end_hour, end_minute = map(int, end.split(":"))

    # Convert current time to minutes since midnight
    current_minutes = local_parts.hour * 60 + local_parts.minute
    start_minutes = start_hour * 60 + start_minute
    end_minutes = end_hour * 60 + end_minute

    # Check if current time is within working hours
    if start_minutes <= end_minutes:
        # Normal case: start and end are on the same day
        return start_minutes <= current_minutes <= end_minutes
    else:
        # Edge case: working hours span midnight (e.g., 22:00 to 06:00)
        return current_minutes >= start_minutes or current_minutes <= end_minutes


def in_working_hours_london(
    date: datetime,
    start: str = DEFAULT_WORKING_HOURS["start"],
    end: str = DEFAULT_WORKING_HOURS["end"],
) -> bool:
    """
    Check if a timestamp falls within London working hours (convenience function).

    Args:
        date: The date to check
        start: Start time in HH:MM format (defaults to '09:00')
        end: End time in HH:MM format (defaults to '17:30')

    Returns:
        True if the timestamp is within London working hours

    Raises:
        ValueError: If inputs are invalid
    """
    return in_working_hours(date, "Europe/London", start, end)


def is_working_day(
    date: datetime,
    tz: str = DEFAULT_TIMEZONE,
    working_days: WorkingDays = None,
) -> bool:
    """
    Check if a date falls on a working day.

    Args:
        date: The date to check
        tz: The timezone identifier (defaults to Europe/London)
        working_days: List of working days (0=Monday, 1=Tuesday, etc.) (defaults to Monday-Friday)

    Returns:
        True if the date is a working day

    Raises:
        ValueError: If inputs are invalid
    """
    if working_days is None:
        working_days = list(DEFAULT_WORKING_DAYS)

    validate_date(date)
    validate_timezone(tz)
    validate_working_days(working_days)
    validate_platform_timezone(tz)

    # Get timezone-local date to determine the correct day of week
    local_parts = to_timezone_parts(date, tz)

    # Create a date in the local timezone to get the correct day of week
    # Note: Python's weekday() returns 0=Monday, 1=Tuesday, etc.
    from datetime import date as date_class
    local_date = date_class(
        local_parts.year, local_parts.month, local_parts.day)
    day_of_week = local_date.weekday()  # 0=Monday, 1=Tuesday, ..., 6=Sunday

    return day_of_week in working_days

File: src/working-hours.ts
/**
 * Working hours and business day utilities
 */

import type { WorkingDays } from "./types.js";
import {
  validateDate,
  validateTimezone,
  validatePlatformTimezone,
} from "./utils/validation.js";
import { validateWorkingDays } from "./validator.js";
import { toTimezoneParts } from "./time-converter.js";
import { timezoneDetector } from "./timezone-detector.js";
import { DEFAULT_WORKING_HOURS, DEFAULT_WORKING_DAYS } from "./constants.js";

/**
 * Check if a timestamp falls within working hours for a given timezone
 *
 * Converts the input timestamp to the specified timezone's local time and checks
 * if it falls within the defined working hours range. Handles timezone-aware
 * calculations correctly, including DST transitions.
 *
 * When no timezone is provided, automatically detects the user's timezone.
 *
 * @param date - The date to check (must be a valid Date object)
 * @param timezone - IANA timezone identifier (optional, auto-detects if omitted)
 * @param start - Start time in HH:MM format (defaults to '09:00')
 * @param end - End time in HH:MM format (defaults to '17:30')
 * @returns `true` if the timestamp is within working hours in the specified or detected timezone
 *
 * @throws {Error} If date is invalid (NaN) or outside supported range (1970-2100)
 * @throws {Error} If timezone is not supported or unavailable on platform
 * @throws {Error} If time format is invalid (must be HH:MM format, e.g., '09:00')
 * @throws {TimezoneDetectionError} If timezone detection fails and fallback is invalid
 *
 * @example
 * ```typescript
 * const utcTime = new Date('2024-07-15T13:00:00Z');  // 1 PM UTC
 *
 * // With explicit timezone and default working hours (09:00-17:30)
 * console.log(inWorkingHours(utcTime, 'Europe/London'));     // true (14:00 BST)
 * console.log(inWorkingHours(utcTime, 'America/New_York'));  // false (09:00 EDT, at start)
 * console.log(inWorkingHours(utcTime, 'Asia/Tokyo'));        // false (22:00 JST, after hours)
 *
 * // With auto-detection (uses user's timezone)
 * console.log(inWorkingHours(utcTime));                      // true/false based on user's timezone
 *
 * // Custom working hours
 * console.log(inWorkingHours(utcTime, 'Europe/London', '08:00', '16:00')); // true
 *
 * // Midnight-spanning hours (22:00-06:00)
 * const nightTime = new Date('2024-07-15T23:00:00Z');  // 11 PM UTC
 * console.log(inWorkingHours(nightTime, 'Europe/London', '22:00', '06:00')); // true (00:00 BST)
 * ```
 */
export function inWorkingHours(
  date: Date,
  timezone?: string,
  start: string = DEFAULT_WORKING_HOURS.start,
  end: string = DEFAULT_WORKING_HOURS.end
): boolean {
  validateDate(date);

  // Validate time strings inline (only used here)
  const validateTimeString = (time: string): void => {
    if (typeof time !== "string") {
      throw new Error("Invalid time: expected string");
    }
    const timeRegex = /^([01][0-9]|2[0-3]):([0-5][0-9])$/;
    if (!timeRegex.test(time)) {
      throw new Error(
        `Invalid time format: ${time}. Expected HH:MM format (e.g., '09:00', '17:30')`
      );
    }
  };

  validateTimeString(start);
  validateTimeString(end);

  // Use auto-detection if no timezone provided
  const effectiveTimezone = timezone ?? timezoneDetector.getDetectedTimezone();
  validateTimezone(effectiveTimezone);

  validatePlatformTimezone(effectiveTimezone);

  // Get timezone-local time components
  const localParts = toTimezoneParts(date, effectiveTimezone);

  // Parse start and end times
  const [startHour, startMinute] = start.split(":").map(Number);
  const [endHour, endMinute] = end.split(":").map(Number);

  // Convert current time to minutes since midnight
  const currentMinutes = localParts.hour * 60 + localParts.minute;
  const startMinutes = startHour! * 60 + startMinute!;
  const endMinutes = endHour! * 60 + endMinute!;

  // Check if current time is within working hours
  if (startMinutes <= endMinutes) {
    // Normal case: start and end are on the same day
    return currentMinutes >= startMinutes && currentMinutes <= endMinutes;
  } else {
    // Edge case: working hours span midnight (e.g., 22:00 to 06:00)
    return currentMinutes >= startMinutes || currentMinutes <= endMinutes;
  }
}

/**
 * Check if a timestamp falls within London working hours (convenience function)
 * @param date - The date to check
 * @param start - Start time in HH:MM format (defaults to '09:00')
 * @param end - End time in HH:MM format (defaults to '17:30')
 * @returns True if the timestamp is within London working hours
 * @throws Error if inputs are invalid
 */
export function inWorkingHoursLondon(
  date: Date,
  start: string = DEFAULT_WORKING_HOURS.start,
  end: string = DEFAULT_WORKING_HOURS.end
): boolean {
  return inWorkingHours(date, "Europe/London", start, end);
}

/**
 * Check if a date falls on a working day
 *
 * When no timezone is provided, automatically detects the user's timezone.
 *
 * @param date - The date to check
 * @param timezone - The timezone identifier (optional, auto-detects if omitted)
 * @param workingDays - Array of working days (0=Sunday, 1=Monday, etc.) (defaults to Monday-Friday)
 * @returns True if the date is a working day
 * @throws Error if inputs are invalid
 * @throws {TimezoneDetectionError} If timezone detection fails and fallback is invalid
 */
export function isWorkingDay(
  date: Date,
  timezone?: string,
  workingDays: WorkingDays = [...DEFAULT_WORKING_DAYS]
): boolean {
  validateDate(date);
  validateWorkingDays(workingDays);

  // Use auto-detection if no timezone provided
  const effectiveTimezone = timezone ?? timezoneDetector.getDetectedTimezone();
  validateTimezone(effectiveTimezone);

  validatePlatformTimezone(effectiveTimezone);

  // Get timezone-local date to determine the correct day of week
  const localParts = toTimezoneParts(date, effectiveTimezone);

  // Create a date in the local timezone to get the correct day of week
  // Note: We use the local parts to construct a date that represents the local date
  const localDate = new Date(
    localParts.year,
    localParts.month - 1,
    localParts.day
  );
  const dayOfWeek = localDate.getDay(); // 0=Sunday, 1=Monday, etc.

  return workingDays.includes(dayOfWeek);
}
